scripts/core/EXT_Exp_SharedContexts.rst

```diff
@@ -0,0 +1,51 @@
+<%
+import re
+from templates import helper as th
+%><%
+    OneApi=tags['$OneApi']
+    x=tags['$x']
+    X=x.upper()
+%>
+:orphan:
+
+.. _ZE_experimental_shared_contexts:
+
+=====================================
+ Shared-Contexts Extension
+=====================================
+
+API
+----
+
+* Enumerations
+
+
+    * ${x}_shared_contexts_exp_version_t
+
+
+* Functions
+
+    * ${x}MemGetAllocPropertiesExp
+    * ${x}DriverGetSharedContextExp
+    * ${x}DriverAddContextToSharedExp
+
+Shared Contexts
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Contexts provide a way to isolate resources in an Level Zero application. Specifically,
```

Comment:
what's the use case here?
We need to be careful to make sure that we do not change the basic paradigm, where resources allocated in one context will be automatically available in another context.

If apps want this kind of behaviour then it needs to allocate shared context from multiple contexts and use this shared context for resource / program /kernel allocation.

Reply:
@MichalMrozek the idea is as you say above "where resources allocated in one context will be automatically available in another context.". The basic idea is to allow users to share resources (previously created or new) among contexts.

Once a shared-context has been created, then any resource previously or later created with one of those contexts is automatically available in the other contexts belonging to the same shared-context.

Whether the new resources are created with the shared context or with one of the contexts ultimately doesnt make a difference, since at the end they are shared among of all them.

Now, if we do as you mention above, that the only resources to be shared are those created with the shared-context, that could cause a problem of data races. Imagine you have two threads, one creating resources with context 0 and other with context 1. At some point, thread 1 creates a shared-context with the two contexts1.

If we do as you say, then only the resources created previously to thread 1 creating the shared context would be shared, but after that, thread 0 would have to use the shared context, but how thread 0 would know when or if a shared-context was created? I think that is overly complicated.

By creating the share-context, the application has already told the driver that wants this kind of behavior and that resources belonging to those contexts would be shared.